crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -0,0 +1,561 @@
+//! Applies the protocol fee to the solution received from the solver.
+//!
+//! Protocol fee is applied by increasing already existing fee determined by the
+//! solver.
+
+use {
+    super::trade::{Fee, Fulfillment, InvalidExecutedAmount},
+    crate::domain::{
+        competition::{
+            order,
+            order::{FeePolicy, Side},
+        },
+        eth,
+    },
+};
+
+impl Fulfillment {
+    pub fn add_protocol_fee(
+        &self,
+        uniform_sell_price: eth::U256,
+        uniform_buy_price: eth::U256,
+    ) -> Result<Self, InvalidExecutedAmount> {
+        let order = self.order().clone();
+        let executed = self.executed();
+        let surplus_fee = match self.raw_fee() {
+            Fee::Static => eth::U256::default(),
+            Fee::Dynamic(fee) => fee.0,
+        };
+
+        let protocol_fee = order
+            .fee_policies
+            .iter()
+            .map(|fee_policy| {
+                match (fee_policy, order.side) {
```

Comment:
nit: move this into a separate method (on self) to make the main `with_protocol_function` easy to read?

Reply:
refactored